File: client/protocol.js
import * as game from "./game.js"
import * as utils from "./utils.js"
import {extractHeaderInfo, generateHeader} from "./utils.js";

export let conn = new WebSocket("ws://localhost:8001");

const PacketType = {
    NEW: 0b0001,
    JOIN: 0b0010,
    QUIT: 0b0011,
    START: 0b0100,
    RESET: 0b0101,
    MOVE: 0b0110,
    UPDATE: 0b0111,
    REFRESH: 0b1000,

    UNKNOWN_PT: 0b1111
}


function pt2name(pt, isAck, isErr, errCode) {
    let type = isErr? `_ERR (${errCode})` : isAck ? "_ACK" : "";
    let arr = Object.entries(PacketType).filter(e => e[1] === pt).map(o => o[0])
    if (arr.length === 0) return "UNKNOWN [" + pt + "]";
    else return arr[0] + type;
}

function send(pt, data = [], isAck = false, isErr = false, code = 0) {
    let arr = [].concat(generateHeader(pt, isAck, isErr, code), data)
    conn.send(new Uint8Array(arr).buffer);
    console.debug("[protocol] send(): " + arr);
}

function onmessage(e) {
    let arr = new Uint8Array(e.data);
    let [pt, isAck, isErr, code] = extractHeaderInfo(arr[0]);
    console.debug("[protocol] onmessage(): recv message (" + pt2name(pt, isAck, isErr, code) + ")");
    switch (pt) {
        case PacketType.NEW: return recv_NEW_ACK(arr);
        case PacketType.JOIN: return isErr? recv_JOIN_ERR(code) : recv_JOIN_ACK(arr);
        case PacketType.QUIT: return recv_QUIT_ACK();
        case PacketType.START: return recv_START();
        case PacketType.RESET: return recv_RESET();
        case PacketType.MOVE: return isErr? recv_MOVE_ERR(code) : recv_MOVE_ACK(arr);
        case PacketType.UPDATE: return recv_UPDATE(arr);
        case PacketType.REFRESH: return isErr? recv_REFRESH_ERR(code) : recv_REFRESH_ACK(arr);
        case PacketType.UNKNOWN_PT: throw Error(`Server received unknown packet: ${pt2name(arr[1])}`);
        default: throw Error(`Received unknown packet: ${pt2name(arr[0])}`);
    }

}

export function init() {
    conn = new WebSocket("ws://localhost:8001");
    conn.binaryType = "arraybuffer";
    conn.onopen = () => {
        console.debug("[protocol] onopen(): conn opened");
        send_REFRESH_REQ();
    }
    conn.onmessage = onmessage;
    conn.onclose = () => {
        conn.close();
        console.debug("[protocol] onclose(): conn close");
    };
    conn.onerror = () => console.error("[protocol] onerror(): error occurred (" + e.type + ")");
    console.debug("[protocol] init(): conn created");
}

// ########## NEW ##########

export function send_NEW_REQ() {
    send(PacketType.NEW);
    console.info("[protocol] send_NEW_REQ()");
}

function recv_NEW_ACK(arr) {
    let gameKey = utils.bytes2str(arr.slice(1, 5));
    let playerKey = utils.bytes2str(arr.slice(5));
    game.handleNew(gameKey, playerKey);
}

// ########## JOIN #########

export function send_JOIN_REQ(gameKey) {
    send(PacketType.JOIN, [...utils.str2bytes(gameKey)]);
    console.info("[protocol] send_JOIN_REQ(): " + gameKey);
}

function recv_JOIN_ACK(arr) {
    let playerKey = utils.bytes2str(arr.slice(1));
    game.handleJoin(playerKey);
}

function recv_JOIN_ERR(errCode) {
    if (errCode === 1) game.handleGameDoesntExists();
    else if (errCode === 2) game.handleGameAlreadyFull();
}

// ########## QUIT #########

export function send_QUIT_REQ() {
    send(PacketType.QUIT, [...utils.getPKeyBytes()]);
    console.info("[protocol] send_QUIT_REQ()");
}

function recv_QUIT_ACK() {
    localStorage.clear();
    game.switchMode(game.Modes.IDLE);
}


// ########## START ########

function recv_START() {
    console.log("Start game");
    game.switchMode(game.Modes.P_TURN);
}

// ########## RESET ########

function recv_RESET() {
    console.log("Reset game");
    game.switchMode(game.Modes.WAIT);
}

// ########## MOVE #########

export function send_MOVE_REQ(a, b) {
    send(PacketType.MOVE, [(a << 4) | b, ...utils.getPKeyBytes()]);
    console.info("[protocol] send_MOVE_REQ(): ["+ a + ", " + b + "]");
}

function recv_MOVE_ACK(arr) {
    let [aType, bType, isEnd, isCh, score, a, b] = utils.parseMoveRes(arr);
    game.showTiles(a, b, aType, bType, isCh);

    document.getElementById("p_score").innerText = score.toString()
    if (isEnd) game.handleGameEnded();
    if (isCh) game.switchMode(game.Modes.O_TURN);
}

function recv_MOVE_ERR(errCode) {
    if (errCode === 1) utils.showMessage("It's not your move!");
    else if (errCode === 2) utils.showMessage("One of the tiles was already picked!");
}

// ########## UPDATE #######

function recv_UPDATE(arr) {
    let [aType, bType, isEnd, isCh, score, a, b] = utils.parseMoveRes(arr);
    game.showTiles(a, b, aType, bType, isCh);

    document.getElementById("o_score").innerText = score.toString()
    if (isEnd) game.handleGameEnded();
    if (isCh) game.switchMode(game.Modes.P_TURN);
}

// ########## REFRESH ######

function send_REFRESH_REQ() {
    send(PacketType.REFRESH, [...utils.getPKeyBytes()]);
    console.info("[protocol] send_REFRESH_REQ()");
}

function recv_REFRESH_ACK(arr) {
    let is_end = (arr[0] & 0b10) !== 0;
    let is_p_turn = (arr[0] & 0b1) !== 0;
    let board = arr.slice(3);

    document.getElementById("p_score").innerText = arr[1].toString()
    document.getElementById("o_score").innerText = arr[2].toString()

    board.map((elem, i) => {
        let aType = elem >> 4;
        let bType = elem & 0b1111;
        game.showTiles(i * 2, i * 2 + 1, aType, bType, false);
    });

    if (is_end) game.handleGameEnded();
    else game.switchMode(is_p_turn? game.Modes.P_TURN : game.Modes.O_TURN);
}

function recv_REFRESH_ERR(code) {
    if (code === 1) {
        localStorage.clear();
        game.switchMode(game.Modes.IDLE);
        utils.showMessage("Your game no longer exists!");
    }
}

File: server/protocol.py
import memory
from packet import PacketType, extract_header_info, generate_header, str2bytes


def process_NEW_REQ(ws, games: memory.Games) -> bytes:
    print(f"NEW|REQ")
    game_key, player_key = games.createGame(ws)
    print(f"\tCreated new game \"{game_key}\" with player \"{player_key}\"")
    return generate_header(PacketType.NEW) + str2bytes(game_key) + str2bytes(player_key)


async def process_JOIN_REQ(ws, game_key: str, games: memory.Games) -> bytes:
    print(f"JOIN|REQ: game=\"{game_key}\"")
    player_key, r = games.addPlayer(game_key, ws)
    if r == 0:
        print(f"\tAdded player \"{player_key}\" to game")
        print(f"\tSending message to first player to start a game")
        await games.players_ws[games.games[game_key].active].send(generate_header(PacketType.START))
        return generate_header(PacketType.JOIN) + str2bytes(player_key)
    elif r == 1:
        print("\tRequested game doesn't exist")
        return generate_header(PacketType.JOIN, is_err=True, code=1)
    elif r == 2:
        print("\tRequested game is already full")
        return generate_header(PacketType.JOIN, is_err=True, code=2)


async def process_QUIT_REQ(player_key: str, games: memory.Games) -> bytes:
    print(f"QUIT|REQ: player=\"{player_key}\"")
    left_player = games.removePlayer(player_key)
    if left_player is not None:
        await games.players_ws[left_player].send(generate_header(PacketType.RESET))
    return generate_header(PacketType.QUIT)


async def process_MOVE_REQ(moves: int, player_key: str, games: memory.Games) -> bytes:
    a = moves >> 4
    b = moves & 0b1111
    print(f"MOVE|REQ: player=\"{player_key}\", {a=}, {b=}")
    res = games.move(player_key, a, b)
    if type(res) == int:
        return generate_header(PacketType.MOVE, is_err=True, code=res)
    b_a, b_b, is_end, is_ch, score = res
    code = ((is_end << 1) | is_ch)
    pt = generate_header(PacketType.UPDATE, code=code) + bytes([((b_a << 4) | b_b), score, moves])
    await games.players_ws[games.getOtherPlayer(player_key)].send(pt)
    return generate_header(PacketType.MOVE, code=code) + bytes([((b_a << 4) | b_b), score, moves])


def process_REFRESH(ws, player_key: str, games: memory.Games) -> bytes:
    print(f"REFRESH|REQ: player=\"{player_key}\"")
    if player_key not in games.players:
        return generate_header(PacketType.REFRESH, is_err=True, code=1)

    game = games.games[games.players[player_key]]
    games.players_ws[player_key] = ws
    code, p_score, o_score, board = game.getStatus(player_key)

    return generate_header(PacketType.REFRESH, code=code) + bytes([p_score, o_score] + board)


async def generate_response(websocket, req: bytes, games: memory.Games) -> bytes:
    pt, is_ack, is_err, code = extract_header_info(req[0])

    if pt == PacketType.NEW:
        return process_NEW_REQ(websocket, games)
    elif pt == PacketType.JOIN:
        return await process_JOIN_REQ(websocket, req[1:].decode("utf-8"), games)
    elif pt == PacketType.QUIT:
        return await process_QUIT_REQ(req[1:].decode("utf-8"), games)
    elif pt == PacketType.MOVE:
        return await process_MOVE_REQ(req[1], req[2:].decode("utf-8"), games)
    elif pt == PacketType.REFRESH:
        return process_REFRESH(websocket, req[1:].decode("utf-8"), games)
    else:
        print(f"Unknown packet type: {pt}")
        return bytes([PacketType.UNKNOWN_PT, pt])


async def process_message(websocket, req: bytes, games: memory.Games):
    res = await generate_response(websocket, req, games)
    res_repr = "|".join([f"{r:08b}" for r in res])
    print(f"\tSending response: {res_repr}")
    await websocket.send(res)


def main():
    pass


if __name__ == '__main__':
    main()
